pkg/controller/cluster/agent/ports.go

```diff
@@ -0,0 +1,253 @@
+		},
+		Data: map[string]string{
+			"range":          portRange,
+			"allocatedPorts": "",
+		},
+		BinaryData: map[string][]byte{
+			"snapshotData": []byte(""),
+		},
+	}
+}
+
+func (a *PortAllocator) getOrCreate(ctx context.Context, client ctrlruntimeclient.Client, configmap *v1.ConfigMap, portRange string) error {
+	nn := types.NamespacedName{
+		Name:      configmap.Name,
+		Namespace: configmap.Namespace,
+	}
+	if err := client.Get(ctx, nn, configmap); err != nil {
+		if apierrors.IsNotFound(err) {
+			// creating the configMap for the first time
+			configmap = a.cm(configmap.Name, configmap.Namespace, portRange)
+			if err := client.Create(ctx, configmap); err != nil {
+				return fmt.Errorf("failed to create port range configmap: %v", err)
+			}
+		} else {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (a *PortAllocator) AllocateWebhookPort(ctx context.Context, cfg *Config) (*int, error) {
+	return a.allocatePort(ctx, cfg, a.WebhookCM, a.WebhookPortRange)
+}
+
+func (a *PortAllocator) DeallocateWebhookPort(ctx context.Context, client ctrlruntimeclient.Client, clusterName, clusterNamespace string, webhookPort int) error {
+	return a.deallocatePort(ctx, client, clusterName, clusterNamespace, a.WebhookCM, webhookPort, a.WebhookPortRange)
+}
+
+func (a *PortAllocator) AllocateKubeletPort(ctx context.Context, cfg *Config) (*int, error) {
+	return a.allocatePort(ctx, cfg, a.KubeletCM, a.KubeletPortRange)
+}
+
+func (a *PortAllocator) DeallocateKubeletPort(ctx context.Context, client ctrlruntimeclient.Client, clusterName, clusterNamespace string, kubeletPort int) error {
+	return a.deallocatePort(ctx, client, clusterName, clusterNamespace, a.KubeletCM, kubeletPort, a.KubeletPortRange)
+}
+
+// allocatePort will assign port to the cluster from a port Range configured for k3k
+func (a *PortAllocator) allocatePort(ctx context.Context, cfg *Config, configMap *v1.ConfigMap, portRange string) (*int, error) {
+	// get configMap first to avoid conflicts
+	if err := a.getOrCreate(ctx, cfg.client, configMap, portRange); err != nil {
+		return nil, err
+	}
+
+	clusterNameNamespace := cfg.cluster.Name + "-" + cfg.cluster.Namespace
+
+	portsMap, err := parsePortMap(configMap.Data["allocatedPorts"])
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := portsMap[clusterNameNamespace]; ok {
+		return ptr.To(portsMap[clusterNameNamespace]), nil
+	}
+	// allocate a new port and save the snapshot
+	snapshot := core.RangeAllocation{
+		Range: configMap.Data["range"],
+		Data:  configMap.BinaryData["snapshotData"],
+	}
+
+	pa, err := portallocator.NewFromSnapshot(&snapshot)
+	if err != nil {
+		return nil, err
+	}
+
+	next, err := pa.AllocateNext()
+	if err != nil {
+		return nil, err
+	}
+
+	portsMap[clusterNameNamespace] = next
+
+	if err := saveSnapshot(pa, &snapshot, configMap, portsMap); err != nil {
+		return nil, err
+	}
+
+	if err := cfg.client.Update(ctx, configMap); err != nil {
+		return nil, err
+	}
+
+	return ptr.To(next), nil
```

Comment:
Is there a reason on returning this as a pointer?

Reply:
Yes, but we are not checking the nil in any case (I guess it cannot be nil, without the err to be nil). Sometimes it's useful to return the three state, but we are anyway de-referencing the pointer. I think it's safer to just return the zero value.

```go
kubeletPort, err = c.PortAllocator.AllocateKubeletPort(ctx, config)
if err != nil {
	return err
}

cluster.Status.KubeletPort = *kubeletPort

webhookPort, err = c.PortAllocator.AllocateWebhookPort(ctx, config)
if err != nil {
	return err
}

cluster.Status.WebhookPort = *webhookPort
```